plugins/goodix-tp/fu-goodixtp-brlb.c

```diff
@@ -0,0 +1,519 @@
+/*
+ * Copyright (C) 2020 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-goodixtp-common.h"
+
+#define FW_HEADER_SIZE	      512
+#define FW_SUBSYS_INFO_OFFSET 42
+#define FW_SUBSYS_INFO_SIZE   10
+
+struct fw_subsys_info {
+	guint8 type;
+	guint size;
+	guint flash_addr;
+	guint8 *data;
+};
+
+#pragma pack(1)
+struct firmware_summary {
+	guint size;
+	guint checksum;
+	guint8 hw_pid[6];
+	guint8 hw_vid[3];
+	guint8 fw_pid[8];
+	guint8 fw_vid[4];
+	guint8 subsys_num;
+	guint8 chip_type;
+	guint8 protocol_ver;
+	guint8 bus_type;
+	guint8 flash_protect;
+	guint8 reserved[8];
+	struct fw_subsys_info subsys[47];
+};
+#pragma pack()
+
+struct goodix_update_ctrl_t {
+	guint8 *config_data;
+	guint32 config_size;
+	gboolean has_config;
+	gint bin_version;
+	struct firmware_summary fw_summary;
+};
+
+static struct goodix_update_ctrl_t brlb_update_ctrl;
+
+static gboolean
+read_pkg(FuDevice *device, guint32 addr, guint8 *buf, guint32 len, GError **error)
+{
+	guint8 HidBuf[PACKAGE_LEN] = {0};
```

Comment:
can you use the .rs helpers to build this? e.g. https://github.com/fwupd/fwupd/blob/main/plugins/synaptics-prometheus/fu-synaprom.rs would generate you a *lot* of this code for free.

Reply:
I added some more docs here: https://github.com/fwupd/fwupd/blob/main/docs/tutorial.md#the-rustgen-helper